DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/AddDelegateToGroupController.cs

```diff
@@ -0,0 +1,151 @@
+    public class AddDelegateToGroupController : Controller
+    {
+        private const string AddDelegateToGroupCookieName = "AddDelegateToGroupFilter";
+        private readonly CentreCustomPromptHelper centreCustomPromptHelper;
+        private readonly IClockService clockService;
+        private readonly IGroupsDataService groupsDataService;
+        private readonly IGroupsService groupsService;
+        private readonly IJobGroupsDataService jobGroupsDataService;
+        private readonly IUserDataService userDataService;
+
+        public AddDelegateToGroupController(
+            IJobGroupsDataService jobGroupsDataService,
+            IUserDataService userDataService,
+            CentreCustomPromptHelper centreCustomPromptHelper,
+            IGroupsService groupsService,
+            IGroupsDataService groupsDataService,
+            IClockService clockService
+        )
+        {
+            this.centreCustomPromptHelper = centreCustomPromptHelper;
+            this.jobGroupsDataService = jobGroupsDataService;
+            this.userDataService = userDataService;
+            this.groupsService = groupsService;
+            this.groupsDataService = groupsDataService;
+            this.clockService = clockService;
+        }
+
+        [Route("{groupId:int}")]
+        public IActionResult Index(
+            int groupId = 1,
+            string? searchString = null,
+            string? sortBy = null,
+            string sortDirection = BaseSearchablePageViewModel.Ascending,
+            string? filterBy = null,
+            string? filterValue = null,
+            int page = 1
+        )
+        {
+            sortBy ??= DefaultSortByOptions.Name.PropertyName;
+            filterBy = FilteringHelper.GetFilterBy(
+                filterBy,
+                filterValue,
+                Request,
+                AddDelegateToGroupCookieName,
+                DelegateActiveStatusFilterOptions.IsActive.FilterValue
+            );
+
+            var centreId = User.GetCentreId();
+            var jobGroups = jobGroupsDataService.GetJobGroupsAlphabetical();
+            var customPrompts = centreCustomPromptHelper.GetCustomPromptsForCentre(centreId);
+            var delegateUsers = userDataService.GetDelegatesNotRegisteredForGroupByGroupId(groupId, centreId);
+            var groupName = userDataService.GetGroupNameById(groupId);
+
+            var model = new AddDelegateToGroupViewModel(
+                delegateUsers,
+                jobGroups,
+                customPrompts,
+                page,
+                groupId,
+                groupName!,
+                searchString,
+                sortBy,
+                sortDirection,
+                filterBy
+            );
+
+            Response.UpdateOrDeleteFilterCookie(AddDelegateToGroupCookieName, filterBy);
+            return View(model);
+        }
+
+        [Route("DelegateAddedToGroupConfirmation/{groupId:int}/{delegateId:int}")]

```

Comment:
I would of expected a `[POST]` request to this endpoint as opposed to this `[GET]` - for the main reason this endpoint performs actions on a cadidates data and a `GET` endpoint is too easy to `replay` by accident.
I also wonder if the delegate should be validated:
- to ensure it is not already part of the group
- that the delegate does actually belong to the corresponding centre.

Reply:
This has now been refactored.